Aimed Fireball Spells!

GunHog opened the pull request:
:cl: Gun Hog
add: The Space Wizard Federation has finally begun teaching its pupils how to properly aim the Fireball spell!
tweak: Fireball spells, including the Devil version, are now aimed via mouse click. It is recommended to avoid firing at anything too close, as the user can still blow himself up!
/:cl:
- Ported the Fireball spells over to the Click-Intercept system, as was done with the alien neurotoxin spit ability.
- Moved some vars and code to the /proc_holder level.

Fireball On/Off sprites by @WJohn!

**_As this is a major balance change to Wizards, Devils, etc, I expect a moderate amount of controversy. I must ask that only useful feedback be posted, as it is important that I do not miss any code-related comments._**

Merge token: https://github.com/tgstation/tgstation/pull/18548

Known Issues:
- ~~Simple animal wizards cannot use Fireball.~~
- Fast moving Wizards can run into their own fireballs and explode. (I might leave this in as "balance")
- ~~Cannot disable Fireball while stunned or dead.~~

![fireballtest](https://cloud.githubusercontent.com/assets/4955510/16543531/4aaee032-409e-11e6-96b6-e8175bc5880e.gif)

Commit:
Fireball Spell Changes

- Changes the fireball spell to be click-to-fire.
- Converted Devil and Wand fireballs as well.
- Moved some code to the /proc_holder level to work across the /alien
and /spell branches.

Commit:
Merge branch 'tgstation/master' into Better_Fireballs

# Conflicts:
#	code/game/gamemodes/shadowling/shadowling_abilities.dm
#	icons/mob/actions.dmi

coiax commented:
What about the fireball wand and staff?

lzimann commented:
those work like a gun, firing a projectile where you click... I assume they won't be affected by this

GunHog commented:
They are aimed as well, I tested the wand and found that they work the same way as fireball now does.

ChangelingRain commented on code/modules/spells/spell_types/wizard.dm:
```diff
@@ -261,17 +261,65 @@
 	range = 20
 	cooldown_min = 20 //10 deciseconds reduction per rank
-
+	var/fireball_type = /obj/item/projectile/magic/fireball
```
this is probably a nerf because this won't check around it for targets to explode on and is thus, with the changed aiming, way harder to hit with

GunHog replied:
Am I using the wrong path?

GunHog replied:
I fixed that now, it will seek out next tile projectiles as expected, making this an even **more dangerous spell!** I can smell the burning salt now!

Incoming5643 commented:
The old fireball projectile used in the spell had a special function where every time it entered a tile it checked the immediate 9 tiles surrounding it for any valid targets. If it found one it would instantly move to that target and explode.

Unless you change the gun projectile to have this behavior, it's a pretty big nerf.

GunHog commented:
I will see what I can do about that, then.

Zilenan91 commented:
Yeah if it doesn't have that feature then it's making fireball as a spell completely useless

GunHog commented:
Fireball in this code does require a direct hit to work. I am not exactly sure how to preserve that, or what I did to make it stop working, but I will resolve it.

Commit:
Tweaks and Fixes

- Infernal Fireball is now a projectile and has the right explosion
values.
- Removed non-projectile spell versions of fireball, as they are unused
- Fixed fireball not seeking adjacent targets.
- Fixed incorrect bolding slashes.

GunHog commented:
Alrighty! It seems that the fireball's (projectile version) adjacent seeking abilities never did work in the first place! That is fixed now.

Infernal fireball, used by Devils, is now also a projectile with its respective values (more fire, less explosion)

coiax commented:
Should put that warning in the changelog.

lzimann commented:
it was always like that

Zilenan91 commented:
Did you check if it kills the wizard that fires it or does it not trigger off of them.

GunHog commented:
It does not trigger on the wizard himself, but it will hit him if he fires at a mob standing too close in the direction he fires.

Zilenan91 commented:
Just like it was before. Excellent.

timkoster1 commented:
If it's 3 tilea wide, wouldn't that pretty much fuck the wiz id he is in a small tunnel like maint?

WJohn commented:
If he doesn't have good aim, yes. But fireball is currently mostly used as a sort of artillery attack because of how dangerous it is to use up-close. This just makes it much less clunky for direct combat.

Supermichael777 commented:
I was toying around with doing this with a one shot gun.

also you aren't a true wizard until you blow your bollocks off with a fireball.

GunHog commented:
> If it's 3 tilea wide, wouldn't that pretty much fuck the wiz id he is in a small tunnel like maint?

Not really! It searches for other mobs to the NE and NW diagonals in its path. If you aim directly down the tunnel, you can manage to fire it without self-destructing. Be sure to watch for mice, though.

Also note that it will not hit lying mobs, so you can use your Magic Missile stun to fire over a target.

Zilenan91 commented:
It should really hit lying mobs. It did that before so that if you KO'd someone in a hallway the fireball would make sure they were dead if you shot it down the hall.

Commit:
Fixes and Tweak

- Fixes Simple Animal wizards
- Fixes Merge Conflict
- Allows swapping to another ranged ability without having to cancel the
first.

GunHog commented:
> It should really hit lying mobs. It did that before so that if you KO'd someone in a hallway the fireball would make sure they were dead if you shot more down the hall.

Based on reading the past code, the projectile checks for adjacent mobs, and ignores them if they are lying down. I see no reason to alter that, as Fireball has a high chance of being a one-hit kill (stun from the explosion along with being on fire or exposed to space). Additional, the Wizard is likely more concerned about attacking the people still standing rather than on the floor!

coiax commented:
If it doesn't hit people lying down, it's a nerf from what it currently does.

Commit:
Merge branch 'tgstation/master' into Better_Fireballs

# Conflicts:
#	icons/mob/actions.dmi

GunHog commented:
> If it doesn't hit people lying down, it's a nerf from what it currently does.

My previous statement was in error. This spell ignores _dead_ mobs, not lying, stunned ones. Conflicts fixed.

Commit:
Fix

Fixes the fireball sometimes missing.

GunHog commented:
Okay, NOW it works as expected.

optimumtact commented on code/modules/projectiles/projectile/magic.dm:
```diff
@@ -24,27 +24,46 @@
 
+/obj/item/projectile/magic/fireball/infernal
+	name = "infernal fireball"
+	exp_heavy = -1
```
this doesn't evaluate to infinite right?

GunHog replied:
I do not believe it does. Nothing odd happened in testing. I am not exactly sure what the difference between -1 and 0 is for explosion code. I am using the same values that were provided in the original spell.

optimumtact replied:
alright, should be fine then, just wanted a quick sanity check.

optimumtact commented on code/modules/spells/spell.dm:
```diff
@@ -1,14 +1,44 @@
+		user << "<span class='warning'><b>[user.ranged_ability.name]</b> has been replaced by <b>[name]</b>."
+		user.ranged_ability.remove_ranged_ability(user)
+		return 1 //1 for failed, 0 for passed.
+	user.next_click = world.time + 6
```
define pls

GunHog replied:
I will have to make a new one, because I do not believe one exist for this exact value.

optimumtact replied:
that is fine, just #define WIZSPELL_CLICK_COOLDOWN or something like that

optimumtact commented on code/modules/spells/spell.dm:
```diff
@@ -1,14 +1,44 @@
+	if(user.ranged_ability && user.ranged_ability != src)
+		user << "<span class='warning'><b>[user.ranged_ability.name]</b> has been replaced by <b>[name]</b>."
+		user.ranged_ability.remove_ranged_ability(user)
+		return 1 //1 for failed, 0 for passed.
```
use TRUE/FALSE defines please

optimumtact commented:
Almost ready to go :+1:

Commit:
Fixes and tweak

- Uses TRUE/FALSE defines instead of 0 and 1 for boolean variables.
- Ranged abilities now remove themselves tries to use an ability in a
disabled state. (Such as being knocked out)
- Define for the click cooldown.

GunHog commented:
Fireball and Alien spit disable themselves if you try to use them while knocked out or are otherwise prevented from using them, such as a spell being on cooldown (so you can do other things again).

WJohn commented:
Perfect, it's been bugging me that neurotoxin wouldn't toggle off automatically once I'd run out of plasma. Thanks for fixing that :+1:

GunHog commented:
> Perfect, it's been bugging me that neurotoxin wouldn't toggle off automatically once I'd run out of plasma. Thanks for fixing that

Actually, I specifically avoided doing that, as I found it annoying in my testing to turn it back on. I can change that if you want, though.

optimumtact commented:
Great work and thanks for going the extra mile on the tweaks
